tests/integrations/server/raft_client.rs

```diff
@@ -67,25 +69,109 @@ fn test_batch_raft_fallback() {
     let mut raft_client = get_raft_client(&pool);
     let counter = Arc::new(AtomicUsize::new(0));
 
-    // Try to bind the mock server on a TCP port, and then do test.
-    for i in 0..100 {
-        let kv_service = MockKv(MockKvForRaft(Arc::clone(&counter)));
-        let port = 60000 + i;
-        let mut mock_server = match mock_kv_service(kv_service, "localhost", port) {
+    let service = MockKvForRaft(Arc::clone(&counter));
+    let (mock_server, port) = create_mock_server(service, 60000, 60100).unwrap();
+
+    let addr = format!("localhost:{}", port);
+    (0..100).for_each(|_| {
+        raft_client.send(1, &addr, RaftMessage::new()).unwrap();
+        thread::sleep(time::Duration::from_millis(10));
+        raft_client.flush();
+    });
+
+    assert!(counter.load(Ordering::SeqCst) > 0);
+    pool.shutdown().wait().unwrap();
+    drop(mock_server)
+}
+
+#[test]
+// Test raft_client auto reconnect to servers after connection break.
+fn test_raft_client_reconnect() {
+    #[derive(Clone)]
+    struct MockKvForRaft(Arc<AtomicUsize>);
+
+    impl MockKvService for MockKvForRaft {
+        fn batch_raft(
+            &mut self,
+            ctx: RpcContext<'_>,
+            stream: RequestStream<BatchRaftMessage>,
+            sink: ClientStreamingSink<Done>,
+        ) {
+            let counter = Arc::clone(&self.0);
+            ctx.spawn(
+                stream
+                    .for_each(move |msgs| {
+                        let len = msgs.msgs.len();
+                        counter.fetch_add(len, Ordering::SeqCst);
+                        Ok(())
+                    })
+                    .map_err(|_| drop(sink)),
+            );
+        }
+    }
+
+    let pool = tokio_threadpool::Builder::new().pool_size(1).build();
+    let mut raft_client = get_raft_client(&pool);
+    let counter = Arc::new(AtomicUsize::new(0));
+
+    let service = MockKvForRaft(Arc::clone(&counter));
+    let (mock_server, port) = create_mock_server(service, 50000, 50300).unwrap();
+    let addr = format!("localhost:{}", port);
+
+    // `send` should success.
+    (0..50).for_each(|_| raft_client.send(1, &addr, RaftMessage::new()).unwrap());
+    raft_client.flush();
+    thread::sleep(time::Duration::from_millis(100));
+    assert_eq!(counter.load(Ordering::SeqCst), 50);
+
+    // `send` should fail after the mock server stopped.
+    drop(mock_server);
+    thread::sleep(time::Duration::from_millis(100));
+    assert!((0..100)
+        .map(|_| raft_client.send(1, &addr, RaftMessage::new()))
+        .collect::<Result<(), _>>()
+        .is_err());
+
+    // `send` should success after the mock server restarted.
+    let service = MockKvForRaft(Arc::clone(&counter));
+    let mock_server = create_mock_server_on(service, port);
+    (0..50).for_each(|_| raft_client.send(1, &addr, RaftMessage::new()).unwrap());
+    raft_client.flush();
+    thread::sleep(time::Duration::from_millis(100));
```

Comment:
Is there any better way than sleep? Sleep is not reliable when the CI machine is overload.

Reply:
But it may only be greater than this time. I think it's OK